### PR TITLE
gccrs: add the `-frust-crate=` option to set the crate name

### PR DESCRIPTION
**Pull request by mathstuf**
Signed-off-by: Ben Boeckel <mathstuf@gmail.com>

Fixes: #627

### CODE REVIEW

File: gcc/rust/rust-session-manager.cc [hunk 1]
```diff
@@ -330,6 +328,9 @@ Session::init ()
   // setup backend to GCC GIMPLE
   backend = rust_get_backend ();
 
+  // set the default crate name
+  options.set_crate_name (kDefaultCrateName);
```
Comment (bjorn3) [inline gcc/rust/rust-session-manager.cc:332]: The default in rustc is the source filename without `.rs` or `rust_out` if the source is stdin.

Comment (philberty) [reply gcc/rust/rust-session-manager.cc:332]: Lets raise an issue to fix that default crate name, I think it makes sense to fix that outside of this PR

File: gcc/rust/lang.opt [hunk 2]
```diff
@@ -34,6 +34,10 @@ L
 Rust Joined Separate
 ; Not documented
 
+frust-crate=
+Rust Joined RejectNegative
+-frust-crate=<naem>             Set the crate name for the compilation
```
Comment (philberty) [inline gcc/rust/lang.opt:39]: typo 'name'

File: gcc/rust/rust-session-manager.h [hunk 3]
```diff
@@ -203,6 +204,13 @@ struct CompileOptions
     enable_dump_option (DumpOption::HIR_DUMP);
     enable_dump_option (DumpOption::TYPE_RESOLUTION_DUMP);
   }
+
+  bool set_crate_name (std::string name)
```
Comment (philberty) [inline gcc/rust/rust-session-manager.h:208]: is there a getter for this? Not sure if we need a move here. I would have either done:

```
bool set_crate_name (std::string&& name) { crate_name = std::move(name); }
```

Or copy

```
void set_crate_name(const std::string& name) { crate_name.assign(name); }
```

**Review by philberty — CHANGES_REQUESTED** [review]
I think after fixing that typo its good to go

**Comment by philberty** [comment]
Also as you pointer out until we get the unit-test framework in there are no tests to add.

File: gcc/rust/rust-session-manager.h (hunk 3 above)
Comment (mathstuf) [reply gcc/rust/rust-session-manager.h:208]: It is public already (`CompileOptions` is a `struct`).

File: gcc/rust/rust-session-manager.cc (hunk 1 above)
Comment (mathstuf) [reply gcc/rust/rust-session-manager.cc:332]: Another question: what if the source doesn't have a `.rs` extension?

Comment (bjorn3) [reply gcc/rust/rust-session-manager.cc:332]: https://github.com/rust-lang/rust/blob/758c00ea4088e1ca2714890f00c791c24f001536/compiler/rustc_session/src/output.rs#L48

Rustc first tries `--crate-name` and `#![crate_name]` (if both are used they must match), then tries to take the filename, take the file stem (filename with last extension removed if any. `a.tar.gz` would result in `a.tar` for the file stem, which is not a valid crate name) and replace `-` with `_`. And finally falls back to `rust_out` if the input is not a file. The file extension doesn't matter.

**Review by philberty — APPROVED** [review]

File: gcc/rust/rust-session-manager.cc (hunk 1 above)
Comment (philberty) [reply gcc/rust/rust-session-manager.cc:332]: We still need to figure out how we can handle ```#![crate_name]``` as we don't have a way to feed this back to the compiler driver yet to auto set the output name.

Comment (mathstuf) [reply gcc/rust/rust-session-manager.cc:332]: See #789.

**Pull request merged**